File: data_collection/real_pose_tracking/src/read_franka_ee_pose.cpp
#include <yaml-cpp/yaml.h>

#include <array>
#include <algorithm>
#include <chrono>
#include <cmath>
#include <csignal>
#include <filesystem>
#include <fstream>
#include <future>
#include <iomanip>
#include <iostream>
#include <memory>
#include <optional>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>

#include <franka/exception.h>
#include <franka/robot.h>
#include <franka/robot_state.h>

namespace {

using Clock = std::chrono::system_clock;
using SteadyClock = std::chrono::steady_clock;

volatile std::sig_atomic_t g_stop_requested = 0;

struct Config {
  std::string robot_ip;
  std::string pose_source_field;
  double connect_timeout_s = 10.0;
  double target_hz = 30.0;
  std::optional<int> max_samples;
  std::filesystem::path output_root;
  std::string jsonl_file_name = "franka_ee_pose_deltas.jsonl";
  std::string summary_file_name = "summary.json";
  bool write_jsonl = true;
  bool emit_stdout_records = false;
  bool include_absolute_pose = false;
  bool include_pose_source_field = true;
  bool include_robot_ip = true;
};

struct Matrix4 {
  std::array<double, 16> values{};

  double& operator()(int row, int col) {
    return values[static_cast<std::size_t>(row + 4 * col)];
  }

  double operator()(int row, int col) const {
    return values[static_cast<std::size_t>(row + 4 * col)];
  }
};

struct Quaternion {
  double x = 0.0;
  double y = 0.0;
  double z = 0.0;
  double w = 1.0;
};

struct PoseSample {
  int sample_index = 0;
  double host_timestamp_s = 0.0;
  Matrix4 transform_base_ee;
  std::array<double, 3> position_xyz{};
  Quaternion quaternion_xyzw;
  std::string pose_source_field;
};

std::string JsonEscape(const std::string& input) {
  std::ostringstream oss;
  for (char ch : input) {
    switch (ch) {
      case '\\':
        oss << "\\\\";
        break;
      case '"':
        oss << "\\\"";
        break;
      case '\n':
        oss << "\\n";
        break;
      case '\r':
        oss << "\\r";
        break;
      case '\t':
        oss << "\\t";
        break;
      default:
        oss << ch;
        break;
    }
  }
  return oss.str();
}

template <std::size_t N>
std::string ArrayToJson(const std::array<double, N>& values) {
  std::ostringstream oss;
  oss << "[";
  for (std::size_t i = 0; i < N; ++i) {
    if (i > 0) {
      oss << ", ";
    }
    oss << std::setprecision(15) << values[i];
  }
  oss << "]";
  return oss.str();
}

std::string QuaternionToJson(const Quaternion& q) {
  return ArrayToJson<4>({q.x, q.y, q.z, q.w});
}

std::string Matrix4ToJson(const Matrix4& matrix) {
  std::ostringstream oss;
  oss << "[";
  for (int row = 0; row < 4; ++row) {
    if (row > 0) {
      oss << ", ";
    }
    oss << "[";
    for (int col = 0; col < 4; ++col) {
      if (col > 0) {
        oss << ", ";
      }
      oss << std::setprecision(15) << matrix(row, col);
    }
    oss << "]";
  }
  oss << "]";
  return oss.str();
}

Matrix4 IdentityMatrix4() {
  Matrix4 result;
  result(0, 0) = 1.0;
  result(1, 1) = 1.0;
  result(2, 2) = 1.0;
  result(3, 3) = 1.0;
  return result;
}

Matrix4 Multiply(const Matrix4& a, const Matrix4& b) {
  Matrix4 result{};
  for (int row = 0; row < 4; ++row) {
    for (int col = 0; col < 4; ++col) {
      double sum = 0.0;
      for (int k = 0; k < 4; ++k) {
        sum += a(row, k) * b(k, col);
      }
      result(row, col) = sum;
    }
  }
  return result;
}

Matrix4 InvertTransform(const Matrix4& transform) {
  Matrix4 inverse = IdentityMatrix4();
  for (int row = 0; row < 3; ++row) {
    for (int col = 0; col < 3; ++col) {
      inverse(row, col) = transform(col, row);
    }
  }
  for (int row = 0; row < 3; ++row) {
    double value = 0.0;
    for (int col = 0; col < 3; ++col) {
      value -= inverse(row, col) * transform(col, 3);
    }
    inverse(row, 3) = value;
  }
  return inverse;
}

Quaternion NormalizeQuaternion(const Quaternion& q) {
  const double norm = std::sqrt(q.x * q.x + q.y * q.y + q.z * q.z + q.w * q.w);
  if (norm == 0.0) {
    throw std::runtime_error("Quaternion norm is zero.");
  }
  return Quaternion{q.x / norm, q.y / norm, q.z / norm, q.w / norm};
}

Quaternion RotationMatrixToQuaternion(const Matrix4& transform) {
  const double trace = transform(0, 0) + transform(1, 1) + transform(2, 2);

  Quaternion q;
  if (trace > 0.0) {
    const double s = std::sqrt(trace + 1.0) * 2.0;
    q.w = 0.25 * s;
    q.x = (transform(2, 1) - transform(1, 2)) / s;
    q.y = (transform(0, 2) - transform(2, 0)) / s;
    q.z = (transform(1, 0) - transform(0, 1)) / s;
  } else if (transform(0, 0) > transform(1, 1) && transform(0, 0) > transform(2, 2)) {
    const double s = std::sqrt(1.0 + transform(0, 0) - transform(1, 1) - transform(2, 2)) * 2.0;
    q.w = (transform(2, 1) - transform(1, 2)) / s;
    q.x = 0.25 * s;
    q.y = (transform(0, 1) + transform(1, 0)) / s;
    q.z = (transform(0, 2) + transform(2, 0)) / s;
  } else if (transform(1, 1) > transform(2, 2)) {
    const double s = std::sqrt(1.0 + transform(1, 1) - transform(0, 0) - transform(2, 2)) * 2.0;
    q.w = (transform(0, 2) - transform(2, 0)) / s;
    q.x = (transform(0, 1) + transform(1, 0)) / s;
    q.y = 0.25 * s;
    q.z = (transform(1, 2) + transform(2, 1)) / s;
  } else {
    const double s = std::sqrt(1.0 + transform(2, 2) - transform(0, 0) - transform(1, 1)) * 2.0;
    q.w = (transform(1, 0) - transform(0, 1)) / s;
    q.x = (transform(0, 2) + transform(2, 0)) / s;
    q.y = (transform(1, 2) + transform(2, 1)) / s;
    q.z = 0.25 * s;
  }

  return NormalizeQuaternion(q);
}

std::string CurrentIsoTime() {
  const auto now = Clock::now();
  const auto now_time_t = Clock::to_time_t(now);
  const auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;

  std::tm local_tm{};
#ifdef _WIN32
  localtime_s(&local_tm, &now_time_t);
#else
  localtime_r(&now_time_t, &local_tm);
#endif

  std::ostringstream oss;
  oss << std::put_time(&local_tm, "%Y-%m-%dT%H:%M:%S")
      << "." << std::setw(3) << std::setfill('0') << ms.count();
  return oss.str();
}

double CurrentUnixTimestampSeconds() {
  const auto now = Clock::now();
  return std::chrono::duration<double>(now.time_since_epoch()).count();
}

void HandleSignal(int) {
  g_stop_requested = 1;
}

Config LoadConfig(const std::filesystem::path& config_path) {
  const YAML::Node root = YAML::LoadFile(config_path.string());
  Config config;

  config.robot_ip = root["robot"]["ip"].as<std::string>();
  config.pose_source_field = root["robot"]["pose_source_field"].as<std::string>("O_T_EE");
  config.connect_timeout_s = root["robot"]["connect_timeout_s"].as<double>(10.0);
  config.target_hz = root["sampling"]["target_hz"].as<double>(30.0);
  if (root["sampling"]["max_samples"] && !root["sampling"]["max_samples"].IsNull()) {
    config.max_samples = root["sampling"]["max_samples"].as<int>();
  }

  config.output_root = root["output"]["output_root"].as<std::string>("output");
  config.jsonl_file_name = root["output"]["jsonl_file_name"].as<std::string>("franka_ee_pose_deltas.jsonl");
  config.summary_file_name = root["output"]["summary_file_name"].as<std::string>("summary.json");
  config.write_jsonl = root["output"]["write_jsonl"].as<bool>(true);
  config.emit_stdout_records = root["output"]["emit_stdout_records"].as<bool>(false);

  config.include_absolute_pose = root["recording"]["include_absolute_pose"].as<bool>(false);
  config.include_pose_source_field = root["recording"]["include_pose_source_field"].as<bool>(true);
  config.include_robot_ip = root["recording"]["include_robot_ip"].as<bool>(true);

  return config;
}

std::unique_ptr<franka::Robot> ConnectRobotWithTimeout(
    const std::string& robot_ip,
    double timeout_s) {
  if (timeout_s <= 0.0) {
    throw std::runtime_error("robot.connect_timeout_s must be positive.");
  }

  std::promise<std::unique_ptr<franka::Robot>> promise;
  std::future<std::unique_ptr<franka::Robot>> future = promise.get_future();

  std::thread worker([robot_ip, promise = std::move(promise)]() mutable {
    try {
      auto robot = std::make_unique<franka::Robot>(robot_ip);
      promise.set_value(std::move(robot));
    } catch (...) {
      promise.set_exception(std::current_exception());
    }
  });

  const auto wait_status = future.wait_for(std::chrono::duration<double>(timeout_s));
  if (wait_status == std::future_status::ready) {
    worker.join();
    return future.get();
  }

  worker.detach();
  throw std::runtime_error(
      "Timed out while connecting to Franka robot at " + robot_ip + " after " +
      std::to_string(timeout_s) + " seconds.");
}

Matrix4 TransformFromState(const franka::RobotState& state, const std::string& pose_source_field) {
  Matrix4 transform{};
  const std::array<double, 16>* raw = nullptr;

  if (pose_source_field == "O_T_EE") {
    raw = &state.O_T_EE;
  } else if (pose_source_field == "O_T_EE_c") {
    raw = &state.O_T_EE_c;
  } else {
    throw std::runtime_error("Unsupported pose_source_field: " + pose_source_field);
  }

  transform.values = *raw;
  return transform;
}

PoseSample BuildSample(
    int sample_index,
    double host_timestamp_s,
    const Matrix4& transform_base_ee,
    const std::string& pose_source_field) {
  PoseSample sample;
  sample.sample_index = sample_index;
  sample.host_timestamp_s = host_timestamp_s;
  sample.transform_base_ee = transform_base_ee;
  sample.position_xyz = {transform_base_ee(0, 3), transform_base_ee(1, 3), transform_base_ee(2, 3)};
  sample.quaternion_xyzw = RotationMatrixToQuaternion(transform_base_ee);
  sample.pose_source_field = pose_source_field;
  return sample;
}

void WriteJsonLine(std::ofstream& stream, const std::string& payload) {
  stream << payload << '\n';
  stream.flush();
}

std::string BuildRecordJson(
    const PoseSample& previous_sample,
    const PoseSample& current_sample,
    const Matrix4& delta_transform,
    const std::array<double, 3>& delta_translation,
    const Quaternion& delta_quaternion,
    const Config& config) {
  std::ostringstream oss;
  oss << "{";
  oss << "\"sample_index\": " << current_sample.sample_index << ", ";
  oss << "\"valid\": true, ";
  oss << "\"host_timestamp_s\": " << std::setprecision(15) << current_sample.host_timestamp_s << ", ";
  oss << "\"prev_host_timestamp_s\": " << std::setprecision(15) << previous_sample.host_timestamp_s << ", ";
  oss << "\"curr_host_timestamp_s\": " << std::setprecision(15) << current_sample.host_timestamp_s << ", ";
  oss << "\"delta_transform_prev_to_curr\": " << Matrix4ToJson(delta_transform) << ", ";
  oss << "\"delta_translation_xyz\": " << ArrayToJson<3>(delta_translation) << ", ";
  oss << "\"delta_quaternion_xyzw\": " << QuaternionToJson(delta_quaternion);

  if (config.include_absolute_pose) {
    oss << ", \"prev_position_xyz\": " << ArrayToJson<3>(previous_sample.position_xyz);
    oss << ", \"prev_quaternion_xyzw\": " << QuaternionToJson(previous_sample.quaternion_xyzw);
    oss << ", \"curr_position_xyz\": " << ArrayToJson<3>(current_sample.position_xyz);
    oss << ", \"curr_quaternion_xyzw\": " << QuaternionToJson(current_sample.quaternion_xyzw);
  }

  if (config.include_pose_source_field) {
    oss << ", \"pose_source_field\": \"" << JsonEscape(current_sample.pose_source_field) << "\"";
  }

  if (config.include_robot_ip) {
    oss << ", \"robot_ip\": \"" << JsonEscape(config.robot_ip) << "\"";
  }

  oss << "}";
  return oss.str();
}

std::string BuildSummaryJson(
    const Config& config,
    const std::filesystem::path& output_root,
    const std::filesystem::path& jsonl_path,
    int records_logged,
    int samples_read,
    std::optional<double> first_timestamp_s,
    std::optional<double> last_timestamp_s) {
  std::ostringstream oss;
  oss << "{\n";
  oss << "  \"output_root\": \"" << JsonEscape(output_root.string()) << "\",\n";
  oss << "  \"jsonl_path\": \"" << JsonEscape(jsonl_path.string()) << "\",\n";
  oss << "  \"robot_ip\": \"" << JsonEscape(config.robot_ip) << "\",\n";
  oss << "  \"pose_source_field\": \"" << JsonEscape(config.pose_source_field) << "\",\n";
  oss << "  \"target_hz\": " << std::setprecision(15) << config.target_hz << ",\n";
  oss << "  \"records_logged\": " << records_logged << ",\n";
  oss << "  \"samples_read\": " << samples_read << ",\n";
  oss << "  \"first_host_timestamp_s\": "
      << (first_timestamp_s.has_value() ? std::to_string(*first_timestamp_s) : "null") << ",\n";
  oss << "  \"last_host_timestamp_s\": "
      << (last_timestamp_s.has_value() ? std::to_string(*last_timestamp_s) : "null") << "\n";
  oss << "}\n";
  return oss.str();
}

}  // namespace

int main(int argc, char** argv) {
  std::signal(SIGINT, HandleSignal);

  std::filesystem::path config_path = "config/default.yaml";
  if (argc > 1) {
    config_path = argv[1];
  }

  try {
    const Config config = LoadConfig(config_path);
    if (config.target_hz <= 0.0) {
      throw std::runtime_error("sampling.target_hz must be positive.");
    }

    const std::filesystem::path project_root = std::filesystem::absolute(config_path).parent_path().parent_path();
    const std::filesystem::path output_root = std::filesystem::absolute(project_root / config.output_root);
    const std::filesystem::path jsonl_path = output_root / config.jsonl_file_name;
    const std::filesystem::path summary_path = output_root / config.summary_file_name;

    std::filesystem::create_directories(output_root);

    std::optional<std::ofstream> jsonl_stream;
    if (config.write_jsonl) {
      jsonl_stream.emplace(jsonl_path, std::ios::out | std::ios::trunc);
      if (!jsonl_stream->is_open()) {
        throw std::runtime_error("Cannot open JSONL output file: " + jsonl_path.string());
      }
    }

    auto robot = ConnectRobotWithTimeout(config.robot_ip, config.connect_timeout_s);

    const double dt = 1.0 / config.target_hz;
    int sample_index = 0;
    int records_logged = 0;
    std::optional<PoseSample> previous_sample;
    std::optional<double> first_timestamp_s;
    std::optional<double> last_timestamp_s;

    while (!g_stop_requested) {
      const auto loop_start = SteadyClock::now();
      const double host_timestamp_s = CurrentUnixTimestampSeconds();
      const franka::RobotState state = robot->readOnce();
      const Matrix4 transform_base_ee = TransformFromState(state, config.pose_source_field);

      ++sample_index;
      PoseSample current_sample = BuildSample(
          sample_index,
          host_timestamp_s,
          transform_base_ee,
          config.pose_source_field);

      if (!first_timestamp_s.has_value()) {
        first_timestamp_s = current_sample.host_timestamp_s;
      }
      last_timestamp_s = current_sample.host_timestamp_s;

      if (previous_sample.has_value()) {
        const Matrix4 delta_transform = Multiply(
            InvertTransform(previous_sample->transform_base_ee),
            current_sample.transform_base_ee);
        const std::array<double, 3> delta_translation = {
            delta_transform(0, 3),
            delta_transform(1, 3),
            delta_transform(2, 3),
        };
        const Quaternion delta_quaternion = RotationMatrixToQuaternion(delta_transform);

        const std::string record_json = BuildRecordJson(
            *previous_sample,
            current_sample,
            delta_transform,
            delta_translation,
            delta_quaternion,
            config);

        if (jsonl_stream.has_value()) {
          WriteJsonLine(*jsonl_stream, record_json);
          ++records_logged;
        }
      }

      previous_sample = current_sample;

      if (config.max_samples.has_value() && sample_index >= *config.max_samples) {
        break;
      }

      const auto elapsed = std::chrono::duration<double>(SteadyClock::now() - loop_start).count();
      const double sleep_s = std::max(0.0, dt - elapsed);
      std::this_thread::sleep_for(std::chrono::duration<double>(sleep_s));
    }

    std::ofstream summary_stream(summary_path, std::ios::out | std::ios::trunc);
    summary_stream << BuildSummaryJson(
        config,
        output_root,
        jsonl_path,
        records_logged,
        sample_index,
        first_timestamp_s,
        last_timestamp_s);
    summary_stream.close();
    return 0;
  } catch (const franka::Exception& exc) {
    std::cerr << "[ERROR] libfranka exception: " << exc.what() << "\n";
    return 1;
  } catch (const std::exception& exc) {
    std::cerr << "[ERROR] " << exc.what() << "\n";
    return 1;
  }
}

File: data_collection/tracking/scripts/read_franka_ee_pose.py
from __future__ import annotations

import argparse
import json
import math
import signal
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

import numpy as np
import yaml

try:
    import pylibfranka
except ImportError as exc:  # pragma: no cover
    raise SystemExit(
        "pylibfranka is required for this script. Install it before running."
    ) from exc


SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent
DEFAULT_CONFIG_PATH = PROJECT_ROOT / "config" / "default.yaml"


@dataclass
class PoseSample:
    sample_index: int
    host_timestamp_s: float
    host_time_iso: str
    transform_base_ee: np.ndarray
    position_xyz: np.ndarray
    quaternion_xyzw: np.ndarray
    pose_source_field: str


class StopRequested(Exception):
    pass


class FrankaRobotAdapter:
    def __init__(self, robot_ip: str) -> None:
        self.robot_ip = robot_ip
        self.robot = self._connect(robot_ip)

    def _connect(self, robot_ip: str) -> Any:
        candidate_constructors = [
            ("Robot", lambda cls: cls(robot_ip)),
            ("Robot", lambda cls: cls(ip=robot_ip)),
            ("Robot", lambda cls: cls(robot_ip=robot_ip)),
            ("FrankaRobot", lambda cls: cls(robot_ip)),
            ("FrankaRobot", lambda cls: cls(ip=robot_ip)),
            ("FrankaRobot", lambda cls: cls(robot_ip=robot_ip)),
        ]

        for attr_name, factory in candidate_constructors:
            cls = getattr(pylibfranka, attr_name, None)
            if cls is None:
                continue
            try:
                return factory(cls)
            except TypeError:
                continue

        available = sorted(name for name in dir(pylibfranka) if not name.startswith("_"))
        raise RuntimeError(
            "Unable to construct a Franka robot from pylibfranka. "
            f"Available top-level symbols: {available}"
        )

    def read_state(self) -> Any:
        candidate_methods = [
            "read_once",
            "read_once_state",
            "get_state",
            "robot_state",
            "state",
            "read",
        ]

        for method_name in candidate_methods:
            method = getattr(self.robot, method_name, None)
            if method is None:
                continue
            try:
                return method()
            except TypeError:
                continue

        raise RuntimeError(
            "Unable to read robot state from pylibfranka robot instance. "
            "Tried methods: "
            + ", ".join(candidate_methods)
        )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Read Franka end-effector pose and log adjacent-sample pose deltas."
    )
    parser.add_argument(
        "--config",
        default=str(DEFAULT_CONFIG_PATH),
        help="Path to YAML config file.",
    )
    return parser.parse_args()


def resolve_path(path_value: str | Path, base_dir: Optional[Path] = None) -> Path:
    path = Path(path_value)
    if path.is_absolute():
        return path.resolve()
    root = base_dir if base_dir is not None else PROJECT_ROOT
    return (root / path).resolve()


def load_config(path_value: str | Path) -> dict[str, Any]:
    config_path = resolve_path(path_value)
    if not config_path.exists():
        raise FileNotFoundError(f"Config file not found: {config_path}")
    data = yaml.safe_load(config_path.read_text(encoding="utf-8")) or {}
    if not isinstance(data, dict):
        raise ValueError(f"Config file must contain a YAML mapping: {config_path}")
    return data


def now_iso8601() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.localtime()) + f".{int((time.time() % 1) * 1000):03d}"


def normalize_quaternion_xyzw(quaternion_xyzw: np.ndarray) -> np.ndarray:
    norm = float(np.linalg.norm(quaternion_xyzw))
    if norm == 0.0:
        raise ValueError("Quaternion norm is zero.")
    return quaternion_xyzw / norm


def rotation_matrix_to_quaternion_xyzw(rotation_matrix: np.ndarray) -> np.ndarray:
    trace = float(np.trace(rotation_matrix))
    if trace > 0.0:
        s = math.sqrt(trace + 1.0) * 2.0
        qw = 0.25 * s
        qx = (rotation_matrix[2, 1] - rotation_matrix[1, 2]) / s
        qy = (rotation_matrix[0, 2] - rotation_matrix[2, 0]) / s
        qz = (rotation_matrix[1, 0] - rotation_matrix[0, 1]) / s
    elif rotation_matrix[0, 0] > rotation_matrix[1, 1] and rotation_matrix[0, 0] > rotation_matrix[2, 2]:
        s = math.sqrt(1.0 + rotation_matrix[0, 0] - rotation_matrix[1, 1] - rotation_matrix[2, 2]) * 2.0
        qw = (rotation_matrix[2, 1] - rotation_matrix[1, 2]) / s
        qx = 0.25 * s
        qy = (rotation_matrix[0, 1] + rotation_matrix[1, 0]) / s
        qz = (rotation_matrix[0, 2] + rotation_matrix[2, 0]) / s
    elif rotation_matrix[1, 1] > rotation_matrix[2, 2]:
        s = math.sqrt(1.0 + rotation_matrix[1, 1] - rotation_matrix[0, 0] - rotation_matrix[2, 2]) * 2.0
        qw = (rotation_matrix[0, 2] - rotation_matrix[2, 0]) / s
        qx = (rotation_matrix[0, 1] + rotation_matrix[1, 0]) / s
        qy = 0.25 * s
        qz = (rotation_matrix[1, 2] + rotation_matrix[2, 1]) / s
    else:
        s = math.sqrt(1.0 + rotation_matrix[2, 2] - rotation_matrix[0, 0] - rotation_matrix[1, 1]) * 2.0
        qw = (rotation_matrix[1, 0] - rotation_matrix[0, 1]) / s
        qx = (rotation_matrix[0, 2] + rotation_matrix[2, 0]) / s
        qy = (rotation_matrix[1, 2] + rotation_matrix[2, 1]) / s
        qz = 0.25 * s
    return normalize_quaternion_xyzw(np.array([qx, qy, qz, qw], dtype=np.float64))


def invert_transform(transform: np.ndarray) -> np.ndarray:
    rotation = transform[:3, :3]
    translation = transform[:3, 3]
    inverse = np.eye(4, dtype=np.float64)
    inverse[:3, :3] = rotation.T
    inverse[:3, 3] = -rotation.T @ translation
    return inverse


def parse_transform_4x4(raw_value: Any) -> np.ndarray:
    raw = np.asarray(raw_value, dtype=np.float64).reshape(-1)
    if raw.size != 16:
        raise ValueError("End-effector transform must contain 16 values.")
    # Franka pose arrays are naturally reshaped as column-major homogeneous transforms.
    transform = np.reshape(raw, (4, 4), order="F")
    return transform


def get_pose_transform_and_field(
    robot_state: Any,
    preferred_field: str,
    fallback_fields: list[str],
) -> tuple[np.ndarray, str]:
    candidate_fields = [preferred_field] + [field for field in fallback_fields if field != preferred_field]

    for field_name in candidate_fields:
        if isinstance(robot_state, dict) and field_name in robot_state:
            return parse_transform_4x4(robot_state[field_name]), field_name
        if hasattr(robot_state, field_name):
            return parse_transform_4x4(getattr(robot_state, field_name)), field_name

    if isinstance(robot_state, dict):
        available = sorted(robot_state.keys())
    else:
        available = sorted(name for name in dir(robot_state) if not name.startswith("_"))
    raise RuntimeError(
        "Cannot find end-effector transform field in robot state. "
        f"Tried: {candidate_fields}. Available: {available}"
    )


def build_sample(
    sample_index: int,
    host_timestamp_s: float,
    host_time_iso: str,
    transform_base_ee: np.ndarray,
    pose_source_field: str,
) -> PoseSample:
    rotation = transform_base_ee[:3, :3]
    position = transform_base_ee[:3, 3].astype(np.float64).copy()
    quaternion = rotation_matrix_to_quaternion_xyzw(rotation)
    return PoseSample(
        sample_index=sample_index,
        host_timestamp_s=host_timestamp_s,
        host_time_iso=host_time_iso,
        transform_base_ee=transform_base_ee,
        position_xyz=position,
        quaternion_xyzw=quaternion,
        pose_source_field=pose_source_field,
    )


def append_jsonl(path: Path, payload: dict[str, Any]) -> None:
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(payload) + "\n")


def main() -> None:
    args = parse_args()
    config = load_config(args.config)
    config_path = resolve_path(args.config)

    robot_cfg = config.get("robot", {})
    sampling_cfg = config.get("sampling", {})
    output_cfg = config.get("output", {})
    recording_cfg = config.get("recording", {})

    robot_ip = str(robot_cfg.get("ip", "")).strip()
    if not robot_ip:
        raise ValueError("robot.ip must be set in the config file.")

    preferred_field = str(robot_cfg.get("pose_source_field", "O_T_EE")).strip() or "O_T_EE"
    fallback_fields = [str(item).strip() for item in robot_cfg.get("fallback_pose_source_fields", []) if str(item).strip()]
    target_hz = float(sampling_cfg.get("target_hz", 30.0))
    if target_hz <= 0.0:
        raise ValueError("sampling.target_hz must be positive.")
    dt = 1.0 / target_hz

    max_samples_value = sampling_cfg.get("max_samples", None)
    max_samples = None if max_samples_value in (None, "", "null") else int(max_samples_value)

    output_root = resolve_path(str(output_cfg.get("output_root", "output")), config_path.parent.parent)
    jsonl_file_name = str(output_cfg.get("jsonl_file_name", "franka_ee_pose_deltas.jsonl"))
    summary_file_name = str(output_cfg.get("summary_file_name", "summary.json"))
    write_jsonl = bool(output_cfg.get("write_jsonl", True))
    emit_stdout_records = bool(output_cfg.get("emit_stdout_records", False))

    include_absolute_pose = bool(recording_cfg.get("include_absolute_pose", True))
    include_pose_source_field = bool(recording_cfg.get("include_pose_source_field", True))
    include_robot_ip = bool(recording_cfg.get("include_robot_ip", True))

    output_root.mkdir(parents=True, exist_ok=True)
    jsonl_path = output_root / jsonl_file_name
    summary_path = output_root / summary_file_name

    stop_requested = False

    def handle_sigint(_signum: int, _frame: Any) -> None:
        nonlocal stop_requested
        stop_requested = True

    signal.signal(signal.SIGINT, handle_sigint)

    print(f"[INFO] Connecting to Franka robot at {robot_ip}")
    adapter = FrankaRobotAdapter(robot_ip)
    print(f"[INFO] Logging pose deltas at {target_hz:.2f} Hz")
    print(f"[INFO] Output directory: {output_root}")
    print("[INFO] Press Ctrl+C to stop.")

    previous_sample: Optional[PoseSample] = None
    records_logged = 0
    sample_index = 0
    first_timestamp_s: Optional[float] = None
    last_timestamp_s: Optional[float] = None
    pose_source_field_used: Optional[str] = None

    while not stop_requested:
        loop_start = time.perf_counter()
        host_timestamp_s = time.time()
        host_time_iso = now_iso8601()
        robot_state = adapter.read_state()
        transform_base_ee, pose_source_field = get_pose_transform_and_field(
            robot_state,
            preferred_field=preferred_field,
            fallback_fields=fallback_fields,
        )

        sample_index += 1
        current_sample = build_sample(
            sample_index=sample_index,
            host_timestamp_s=host_timestamp_s,
            host_time_iso=host_time_iso,
            transform_base_ee=transform_base_ee,
            pose_source_field=pose_source_field,
        )

        if first_timestamp_s is None:
            first_timestamp_s = current_sample.host_timestamp_s
        last_timestamp_s = current_sample.host_timestamp_s
        pose_source_field_used = current_sample.pose_source_field

        if previous_sample is not None:
            delta_transform = invert_transform(previous_sample.transform_base_ee) @ current_sample.transform_base_ee
            delta_translation = delta_transform[:3, 3].astype(np.float64)
            delta_quaternion = rotation_matrix_to_quaternion_xyzw(delta_transform[:3, :3])

            record: dict[str, Any] = {
                "sample_index": current_sample.sample_index,
                "valid": True,
                "prev_host_timestamp_s": previous_sample.host_timestamp_s,
                "curr_host_timestamp_s": current_sample.host_timestamp_s,
                "prev_host_time_iso": previous_sample.host_time_iso,
                "curr_host_time_iso": current_sample.host_time_iso,
                "delta_dt_s": current_sample.host_timestamp_s - previous_sample.host_timestamp_s,
                "delta_transform_prev_to_curr": delta_transform.tolist(),
                "delta_translation_xyz": delta_translation.tolist(),
                "delta_quaternion_xyzw": delta_quaternion.tolist(),
            }

            if include_absolute_pose:
                record["prev_position_xyz"] = previous_sample.position_xyz.tolist()
                record["prev_quaternion_xyzw"] = previous_sample.quaternion_xyzw.tolist()
                record["curr_position_xyz"] = current_sample.position_xyz.tolist()
                record["curr_quaternion_xyzw"] = current_sample.quaternion_xyzw.tolist()

            if include_pose_source_field:
                record["pose_source_field"] = current_sample.pose_source_field

            if include_robot_ip:
                record["robot_ip"] = robot_ip

            if write_jsonl:
                append_jsonl(jsonl_path, record)
                records_logged += 1
            if emit_stdout_records:
                print(json.dumps(record), flush=True)
            else:
                print(
                    "[POSE] "
                    f"idx={current_sample.sample_index} "
                    f"dt={record['delta_dt_s']:.4f}s "
                    f"dxyz=({delta_translation[0]:+.4f}, {delta_translation[1]:+.4f}, {delta_translation[2]:+.4f}) "
                    f"dq=({delta_quaternion[0]:+.5f}, {delta_quaternion[1]:+.5f}, {delta_quaternion[2]:+.5f}, {delta_quaternion[3]:+.5f})"
                )

        previous_sample = current_sample

        if max_samples is not None and sample_index >= max_samples:
            break

        elapsed = time.perf_counter() - loop_start
        time.sleep(max(0.0, dt - elapsed))

    summary = {
        "output_root": str(output_root),
        "jsonl_path": str(jsonl_path),
        "robot_ip": robot_ip,
        "target_hz": target_hz,
        "records_logged": records_logged,
        "samples_read": sample_index,
        "pose_source_field": pose_source_field_used,
        "first_host_timestamp_s": first_timestamp_s,
        "last_host_timestamp_s": last_timestamp_s,
    }
    summary_path.write_text(json.dumps(summary, indent=2), encoding="utf-8")
    print(f"[DONE] Summary written to: {summary_path}")


if __name__ == "__main__":
    main()
